server/utilities/transaction.ts

```diff
@@ -0,0 +1,75 @@
+import { StatusCodes } from "http-status-codes";
+import { createCustomError } from "../../core";
+import { getTxById } from "../../src/db/transactions/getTxById";
+import { updateTx } from "../../src/db/transactions/updateTx";
+import { TransactionStatusEnum } from "../schemas/transaction";
+
+export const cancelTransactionAndUpdate = async (
+  queueId: string,
+  walletAddress: string,
+) => {
+  const txData = await getTxById({ queueId });
+
+  console.log("txData", txData);
+
+  if (!txData) {
+    const error = createCustomError(
+      `Transaction not found with queueId ${queueId}`,
+      StatusCodes.NOT_FOUND,
+      "QueueIdNotFound",
+    );
+    throw error;
+  }
+
+  let error = null;
+
+  switch (txData.status) {
+    case TransactionStatusEnum.Cancelled:
+      error = createCustomError(
+        `Transaction already cancelled with queueId ${queueId}`,
+        StatusCodes.BAD_REQUEST,
+        "TransactionAlreadyCancelled",
+      );
+      break;
+    case TransactionStatusEnum.Queued:
+      console.log("inside Mined/Queued");
+      await updateTx({
+        queueId,
+        status: TransactionStatusEnum.Cancelled,
+        txData: {
+          cancelledAt: new Date(),
+        },
+      });
+      break;
+    case TransactionStatusEnum.Mined:
+      console.log("inside Mined");
+      error = createCustomError(
+        `Transaction already mined with queueId ${queueId}`,
+        StatusCodes.BAD_REQUEST,
+        "TransactionAlreadyMined",
+      );
+      break;
+    case TransactionStatusEnum.Processed:
```

Comment:
Won't this method never work if a transaction has already been processed or submitted? Since the worker would still have already picked it up and sent the transaction - but here we're updating the DB state to be out of sync with what actually happened

In the scenario where it's already sent, we may have to sent a replacement transaction with the method @arcoraven mentioned of sending 0 to yourself.

Reply:
In the current implementation, if the transaction is submitted & then we submit cancel, it will have to go through the worker to get executed (in the current implementation). The expectation was that the transaction is taking time and want to cancel it and redo the transaction again.

But if the expectation of `/cancel` is,  we send the cancel transaction with 0 value on the `/cancel` request and not give it a `queueId` then will have to update the flow.

Reply:
@adam-maj , synced with @arcoraven going to make the `/cancel` action synchronous and not go through the worker flow as we need it to get sent on-chain immediately